File: Backend (Springboot)/Annapurna/src/main/java/com/annapurna/service/OrderService.java
package com.annapurna.service;

import com.annapurna.domain.*;
import com.annapurna.dto.OrderDTO;
import com.annapurna.repository.OrderRepository;
import com.annapurna.repository.TiffinRepository;

import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;


import com.lowagie.text.*;
import com.lowagie.text.pdf.*;
import java.io.ByteArrayOutputStream;

@Service
@RequiredArgsConstructor
public class OrderService {

	private final OrderRepository orderRepository;
	private final TiffinRepository tiffinRepository;

	public Order placeOrder(OrderDTO dto, User customer) {
		Tiffin tiffin = tiffinRepository.findById(dto.getTiffinId())
				.orElseThrow(() -> new RuntimeException("Tiffin not found"));

		Order order = Order.builder().customer(customer).tiffin(tiffin).quantity(dto.getQuantity())
				.totalPrice(dto.getQuantity() * tiffin.getPrice()).orderTime(LocalDateTime.now())
				.status(OrderStatus.PENDING).build();

		return orderRepository.save(order);
	}

	public List<Order> getMyOrders(User customer) {
		return orderRepository.findByCustomer(customer);
	}
	

	@Transactional
	public void cancelOrder(Long orderId, User user) {
	    Order order = orderRepository.findById(orderId)
	            .orElseThrow(() -> new RuntimeException("Order not found"));

	    if (!order.getCustomer().getId().equals(user.getId())) {
	        throw new RuntimeException("You are not authorized to cancel this order.");
	    }

	    LocalDateTime now = LocalDateTime.now();
	    if (order.getOrderTime().plusMinutes(5).isBefore(now)) {
	        throw new RuntimeException("Order can only be cancelled within 5 minutes of placing.");
	    }

	    order.setStatus(OrderStatus.CANCELLED);
	    orderRepository.save(order);
	}
	
	
	
	

	public Order getOrderById(Long orderId) {
        return orderRepository.findById(orderId)
                .orElseThrow(() -> new RuntimeException("Order not found with id: " + orderId));
    }
	
	public Order save(Order order) {
	    return orderRepository.save(order);
	}

	
	public void markAsPaid(Long orderId, String razorpayOrderId, String razorpayPaymentId) {
	    Order order = getOrderById(orderId);
	    order.setPaymentStatus(PaymentStatus.PAID);
	    order.setRazorpayOrderId(razorpayOrderId);
	    order.setRazorpayPaymentId(razorpayPaymentId);
	    orderRepository.save(order);
	}



	
	
	
	
	
	
	
	
	
	
	public byte[] generateReceiptPdf(Order order) throws Exception {
	    Document document = new Document();
	    ByteArrayOutputStream out = new ByteArrayOutputStream();

	    PdfWriter.getInstance(document, out);
	    document.open();

	    Font titleFont = new Font(Font.HELVETICA, 20, Font.BOLD);
	    Font normalFont = new Font(Font.HELVETICA, 12);

	    document.add(new Paragraph("Annapurna Tiffins - Order Receipt", titleFont));
	    document.add(new Paragraph(" "));
	    document.add(new Paragraph("Order ID: " + order.getId(), normalFont));
	    document.add(new Paragraph("Customer: " + order.getCustomer().getFullName(), normalFont));
	    document.add(new Paragraph("Tiffin: " + order.getTiffin().getName(), normalFont));
	    document.add(new Paragraph("Quantity: " + order.getQuantity(), normalFont));
	    document.add(new Paragraph("Total Price: ₹" + order.getTotalPrice(), normalFont));
	    document.add(new Paragraph("Order Time: " + order.getOrderTime(), normalFont));
	    document.add(new Paragraph("Payment Status: " + order.getPaymentStatus(), normalFont));
	    document.add(new Paragraph("Razorpay Order ID: " + order.getRazorpayOrderId(), normalFont));

	    document.close();
	    return out.toByteArray();
	}
	
	
	
	
	
	
	
	
	

}

File: Backend (DNET)/Annapurna/Services/OrderService.cs
using Microsoft.EntityFrameworkCore;
using AnnapurnaAPI.Data;
using AnnapurnaAPI.Models;
using AnnapurnaAPI.DTOs;
using iTextSharp.text;
using iTextSharp.text.pdf;
using System.IO;

namespace AnnapurnaAPI.Services
{
    public class OrderService
    {
        private readonly ApplicationDbContext _context;

        public OrderService(ApplicationDbContext context)
        {
            _context = context;
        }

        public async Task<Order> PlaceOrder(OrderDTO dto, User customer)
        {
            var tiffin = await _context.Tiffins.FindAsync(dto.TiffinId);
            if (tiffin == null)
            {
                throw new InvalidOperationException("Tiffin not found");
            }

            var order = new Order
            {
                TiffinId = dto.TiffinId,
                CustomerId = customer.Id,
                Quantity = (int)dto.Quantity,  // Make sure you cast explicitly if it's a long
                TotalPrice = dto.TotalPrice,
                OrderTime = DateTime.UtcNow,
                Status = OrderStatus.PENDING,
                PaymentStatus = PaymentStatus.PENDING
            };

            _context.Orders.Add(order);
            await _context.SaveChangesAsync();

            return order;
        }


        public async Task<List<Order>> GetMyOrders(User customer)
        {
            return await _context.Orders
                .Include(o => o.Tiffin)
                .Where(o => o.CustomerId == customer.Id)
                .OrderByDescending(o => o.OrderTime)
                .ToListAsync();
        }

        public async Task<Order> GetOrderById(long orderId)
        {
            var order = await _context.Orders
                .Include(o => o.Tiffin)
                .Include(o => o.Customer)
                .FirstOrDefaultAsync(o => o.Id == orderId);

            if (order == null)
            {
                throw new InvalidOperationException("Order not found");
            }

            return order;
        }

        public async Task CancelOrder(long orderId, User user)
        {
            var order = await _context.Orders
                .Include(o => o.Customer)
                .FirstOrDefaultAsync(o => o.Id == orderId);

            if (order == null)
            {
                throw new InvalidOperationException("Order not found");
            }

            if (order.CustomerId != user.Id && user.Role != Role.ADMIN)
            {
                throw new InvalidOperationException("Unauthorized to cancel this order");
            }

            if (order.Status == OrderStatus.DELIVERED || order.Status == OrderStatus.CANCELLED)
            {
                throw new InvalidOperationException("Cannot cancel this order");
            }

            order.Status = OrderStatus.CANCELLED;
            await _context.SaveChangesAsync();
        }

        public async Task Save(Order order)
        {
            _context.Orders.Update(order);
            await _context.SaveChangesAsync();
        }

        public async Task MarkAsPaid(long orderId, string razorpayOrderId, string razorpayPaymentId)
        {
            var order = await GetOrderById(orderId);
            order.PaymentStatus = PaymentStatus.PAID;
            order.RazorpayPaymentId = razorpayPaymentId;
            await _context.SaveChangesAsync();
        }

        public byte[] GenerateReceiptPdf(Order order)
        {
            using (MemoryStream ms = new MemoryStream())
            {
                Document document = new Document(PageSize.A4, 25, 25, 30, 30);
                PdfWriter writer = PdfWriter.GetInstance(document, ms);

                document.Open();

                // Add title
                Font titleFont = FontFactory.GetFont(FontFactory.HELVETICA_BOLD, 18);
                Paragraph title = new Paragraph("Order Receipt", titleFont);
                title.Alignment = Element.ALIGN_CENTER;
                document.Add(title);
                document.Add(new Paragraph(" "));

                // Add order details
                Font normalFont = FontFactory.GetFont(FontFactory.HELVETICA, 12);
                document.Add(new Paragraph($"Order ID: {order.Id}", normalFont));
                document.Add(new Paragraph($"Tiffin: {order.Tiffin.Name}", normalFont));
                document.Add(new Paragraph($"Quantity: {order.Quantity}", normalFont));
                document.Add(new Paragraph($"Total Price: ₹{order.TotalPrice}", normalFont));
                document.Add(new Paragraph($"Order Date: {order.OrderTime:dd/MM/yyyy HH:mm}", normalFont));
                document.Add(new Paragraph($"Status: {order.Status}", normalFont));
                document.Add(new Paragraph($"Payment Status: {order.PaymentStatus}", normalFont));

                document.Close();
                return ms.ToArray();
            }
        }
    }
}
